gamer-project/gamer
# Improve Par_EquilibriumIC

## Goal
- To resolve #188 
- To resolve #201 
- To resolve #202 
- To resolve #295 

## Changes
1. Adjust the user interface
    1. Move the part with reading the parameters from the input files into `TestProblem/Hydro/ParticleEquilibriumIC/Init_TestProb_Hydro_ParEqmIC.cpp` like other test problems.
    2. Use `setXXX()` as the interface to set the parameters for the cloud constructors from test problems, and only the necessary parameters are needed to be set.
    3. Support adding analytical external potential (only the Plummer potential now).
    4. Output the relative error of the total enclosed mass by comparing it to the analytical solution instead.
    5. Output the radial distribution and energy distribution for debugging.
    6. Separate the number of bins of the input density profile, input external potential, inner radial profiles, and inner energy distribution.

2. Adjust the calculation
    1.  Unify the processes for density profiles from the table and from the analytical models.
    2. Construct the radial profiles more consistently, from the density profile to the enclosed mass profile, the potential profile , and finally the drho_dPsi profile.
    3. Use the central difference for the differentiation instead of linear regression.
    4. Add the external potential before the usage of potential to be more consistent.
    5. Include the low-binding-energy (high-velocity) distribution function.
    7. Use linear extrapolation for the density profile at r = 0.
    8. Integrate the enclosed mass profile with a higher-order shell mass for the density table.
    9. Use interpolation to sample E from the probability distribution.
    10. Use interpolation to find the drho/dPsi for a given Psi.

3. Adjust the format
    1. Rename some variables and functions.
    2. Remove some no longer used variables and functions.
    3. Add more description.
    4. Adjust the coding styles.


## Verification Tests
1. Compare the distribution to the old branch with the default double-clouds setup. ->The results can roughly match the previous results.
    1. Type = NFW 

    | radial | energy |
    | ---------| ---------- |
    | ![image](https://github.com/gamer-project/gamer/assets/24509818/16ccc6a9-e205-4351-a523-acf9789cbffc) | ![image](https://github.com/gamer-project/gamer/assets/24509818/0b40327e-213d-4e87-aa16-7c8b61c4b5eb) |

    2. Type = Table

   | radial | energy |
   | ---------| ---------- |
   | ![image](https://github.com/gamer-project/gamer/assets/24509818/3ec9aa28-5cb6-4af1-bee5-1062b603026d)| ![image](https://github.com/gamer-project/gamer/assets/24509818/2effcb90-ddb1-4824-aaf5-1da6a9a4ef3c)|

2. Compare the distribution of the Hernquist model to the analytical solutions. -> The results can match analytical solutions.
![image](https://github.com/gamer-project/gamer/assets/24509818/dbca49e7-a28f-4f68-b6df-cf2593725cf0)

3. Check the stability of a single NFW profile with the default setup. -> The results are stable.

| New | Old |
| -------| ------ |
| ![Proj](https://github.com/gamer-project/gamer/assets/24509818/9ca8b23e-2ff3-4ad5-968b-2c1e0fc6d170)| ![Proj](https://github.com/gamer-project/gamer/assets/24509818/b0a3d1c4-c248-4726-a556-caddaf19299d)|
| ![Prof](https://github.com/gamer-project/gamer/assets/24509818/5ca63b39-3de8-41ff-ab5f-ce723298e49e)| ![Prof](https://github.com/gamer-project/gamer/assets/24509818/7d64ad39-16ae-41e9-a216-4e8e609fc029)|
|![fig_ErrorDensityProfile](https://github.com/gamer-project/gamer/assets/24509818/04d4e000-3efd-476f-b10d-97f2e80a4bfa) |![fig_ErrorDensityProfile](https://github.com/gamer-project/gamer/assets/24509818/c07ad9ec-871d-4ed3-a924-c18c9f255dbd) |
4. Check the stability of a single Einasto density profile with the Plummer external potential (constructed with external potential and run with the same external potential). -> The results are stable.

| Projections | Profiles | Error |
| -----------------| ----------- |-------- |
|![Proj](https://github.com/gamer-project/gamer/assets/24509818/8740b335-6248-4f4f-9905-3b33d87295cf)| ![Prof](https://github.com/gamer-project/gamer/assets/24509818/196cd1a4-1a92-46d7-aa4b-0f607a6837de)|![fig_ErrorDensityProfile](https://github.com/gamer-project/gamer/assets/24509818/f88120a5-4e26-4083-be2e-2909cd02a01b)|

## Issue
1. When the E is smaller than the Psi_min ( at r = MaxR ), the dRho/dPsi is approximated using linear extrapolation. However, the result for the distribution function cannot match the analytical solutions, see the Hernquist results above.
